runtime/near-vm-runner/src/imports.rs

```diff
@@ -274,55 +274,150 @@ pub(crate) mod wasmer {
-                    logic.$func( $( $arg_name, )* )
+    trait Wasmer2Type {
+        type Wasmer;
+        fn to_wasmer(self) -> Self::Wasmer;
+        fn ty() -> wasmer_types::Type;
+    }
+    macro_rules! wasmer_types {
+        ($($native:ty as $wasmer:ty => $type_expr:expr;)*) => {
+            $(impl Wasmer2Type for $native {
+                type Wasmer = $wasmer;
+                fn to_wasmer(self) -> $wasmer {
+                    self as _
                 }
+                fn ty() -> wasmer_types::Type {
+                    $type_expr
+                }
+            })*
+        }
+    }
+    wasmer_types! {
+        u32 as i32 => wasmer_types::Type::I32;
+        u64 as i64 => wasmer_types::Type::I64;
+    }
 
-                namespace.insert(stringify!($func), wasmer::Function::new_native_with_env(&store, env.clone(), $func));
-            };
+    macro_rules! return_ty {
+        () => {
+            type Ret = ();
+            fn make_ret() -> () {}
+        };
+        ($($return: ident),*) => {
+            #[repr(C)]
+            struct Ret($(<$return as Wasmer2Type>::Wasmer),*);
+            fn make_ret($($return: $return),*) -> Ret { Ret($($return.to_wasmer()),*) }
         }
-        for_each_available_import!(protocol_version, add_import);
+    }
 
-        import_object.register("env", namespace);
-        import_object
+    impl<'vmlogic, 'vmlogic_refs> Resolver for Wasmer2Imports<'vmlogic, 'vmlogic_refs> {
+        fn resolve(&self, _index: u32, module: &str, field: &str) -> Option<wasmer_engine::Export> {
+            if module != "env" {
+                return None;
+            }
+            if field == "memory" {
+                return Some(wasmer_engine::Export::Memory(self.memory.clone()));
+            }
+
+            macro_rules! add_import {
+                (
+                  $func:ident <
+                    [ $( $arg_name:ident : $arg_type:ident ),* ]
+                    -> [ $( $returns:ident ),* ]
+                  >
+                ) => {
+                    return_ty!($($returns),*);
+
+                    extern "C" fn $func(env: *mut VMLogic<'_>, $( $arg_name: $arg_type ),* )
+                    -> Ret {
+                        // SAFETY: This function should only be executable within `'vmlogic`
+                        // lifetime and so it is safe to dereference the `env` pointer which is
+                        // known to be derived from a valid reference in the first place.
+                        let env = unsafe { &mut *env };
+                        const IS_GAS: bool = str_eq(stringify!($func), "gas");
+                        let _span = if IS_GAS {
+                            None
+                        } else {
+                            Some(tracing::trace_span!(
+                                target: "host-function",
+                                stringify!($func)
+                            ).entered())
+                        };
```

Comment:
Let's move the tracing bit into the `catch_unwind` boundary? Just to limit the amount of code outside to the absolute minimum.